### DISCUSSION TITLE
Pygbag post request

### DISCUSSION THREAD

**Discussion by erasmushaggblom** [pygame-web/pygbag#120] [Q&A] [answered]
Hi,

I am working on a game in pygame, with the plan of deploying it using pygbag. I need to be able to save the game results to a secured server, preferably through sending a POST request to an API. I have been able to integrate that into the Python file, but have not been successful in sending the request using the converted file. I have tried using both urllib and requests in the base python file. My understanding is that I would need to integrate JavaScript into the HTML file and send the request that way. Is this feasible, keeping in mind that I need to trigger the POST request once the game is finished?

**Comment by pmp-p** [comment, marked as answer]
> keeping in mind that I need to trigger the POST request once the game is finished

there's aio.fetch.RequestHandler for that https://github.com/pygame-web/pygbag/blob/6fef0c949ab8dfaa3f403d2e146ded9dd2def16e/pygbag/support/cross/aio/fetch.py#L191

That code was extracted from a scoreboard experiment written by @tank-king but has no test for it right now. Please report back any problem.

for posting on exit it is quite tricky : some platform may allow to close tabs without any warning or callback. You should background that task but as soon as possible / make a "saving" screen animation like most modern games have.


side note: maybe implement some proof of game while game is running and save incrementally game state. because some people could just make arbitrary post request to the server.

**Reply by erasmushaggblom** [reply]
Thank you very much, that worked perfectly! For posting the results on exit, I just placed the request a bit earlier, right before game end is triggered.

**Reply by tank-king** [reply]
Thanks @erasmushaggblom for testing this out. When I wrote that script, it worked fine on my page because I used a custom template file, but it did not work with default versions. I think it has been added later in newer versions, and because it works for you, it is confirmed it works. Thanks.

**Comment by andreagalle** [comment]
Would it be possible to do something similar with databases, like get an asset from a mongodb or make an authenticated request to a service listening via REST API?
